### ISSUE TITLE
Investigate custom data storage solution

### ISSUE THREAD

**Issue by nwalters512** [nwalters512/wildrank-desktop-v2#12]
We currently use Couchbase Lite for all data storage. The benefit of that is that it's a battle-tested library: someone's done all the hard work of squashing bugs, ensuring reliability, and optimizing performance.

However, the downside of that is we are all but forced into using Java for all apps that want to touch the database. Couchbase uses different binary database formats and data storage engines depending on the platform, and as we're forced to interact directly with database files on the flash drive, we have to use their client libraries that use SQLite (I think) as the binary format, so we're pretty much locked into using Java.

The reason this is applicable is because I *hate* desktop Java apps with a passion and I wanted to build a nice [Electron](http://electron.atom.io/) app that is hopefully a lot more usable than what we have now. However, we'd have to bridge from JS to Java somehow, and that makes me cringe just thinking about it. The desktop app isn't exactly user-friendly, and I fear WildRank will fall out of use given that we haven't exactly done a great job of documenting how to use it from end to end.

## Proposal
A good option would be to model after something like [nedb](https://github.com/louischatriot/nedb), where our "database" is just a bunch of JSON-encoded objects in a text file. The obvious benefit there is that UTF-8 files are very obviously portable. We should be able to hold the entire database in memory, as the number of documents we're dealing with is *at most* on the order of 1000.

In fact, all we'd really have to do is build a Java library that conforms to the spec setup by `nedb`. It doesn't have to be particularly fancy in the sense that we don't have to support complex queries; most of our queries will either be fetching objects with a certain key or iterating over all objects of a certain type to generate lists.

Concerns:
* Reliability: USB-OTG adapters aren't exactly the most reliable thing in the world. Couchbase has pretty good corruption protection built in. How hard will it be to make our own work fault-tolerant?
* Performance: This probably won't be a problem given the relatively small number of documents we're dealing with. But it's worth investigating how long it takes to deserialize/query/serialize/etc.

@garwards @smitty6791 @mail929 don't know if any of you are involved enough in WildRank for all of that to mean anything or if you have any thoughts on this idea. Comments welcome.

**Comment by garwards** [comment]
Quick question or two then.. I assume then that the mobile app uses
couchbase as well?

What does the desktop app do that the tablets don't?  I've just never seen
it since I only deal with the tablets.  Is it for entering event data?

**Comment by nkiel** [comment]
It takes the event data off of blue alliance.

**Comment by smitty6791** [comment]
One option you have for data integrity is to run something like a crc32
checksum over the entire data set and append it to the end of the file.
Unfortunately, that is one additional thing you have to write or more
likely borrow from somewhere.  Crc32 is a pretty common checksum routine
and really pretty easy to write in most languages if necessary.  You could
also go with something more complex, but this usually catches everything.
It will increase in processing time, the longer the file gets, but it still
shouldn't be that bad.  All of this may still require custom data
processing when the file is read and loaded.

One question though, since Electron is javascript, can't you just access
the Couchbase java library from the Electron app?

**Comment by garwards** [comment]
If my understanding is correct, this is because the file/encoding format
used by couchbase on android vs windows/mac is different.  But if I think
harder (because I'm notably uncaffeinated right now), you'd need some kind
of neutral API (such as a REST service) to return the data from couchbase
to a JS app, since it wouldn't be able to natively call Java from JS.

Other than using the same service (couchbase etc.) with the same native
file encoding, you need some kind of intermediate file format to transfer
the data.  XML is very verbose, and JSON is less verbose, but either would
serve the purpose.  You need some kind of extraction/population task, but
I'm guessing we're doing that now.  If we're worried about OTG transfers
and losing files, my best suggestion would be to ZIP the items being sent,
so it is one file, with an all-or-nothing type success/failure.  Once it's
on the device, you can extract and handle the files separately.

When I read the phrase 'Javascript database' on the nedb page, I cried a
little.  I would rather see some native code running as a DB to query, and
not a pseudo DB running from JS against a bunch of files.  Memory use/bloat
and performance come to mind immediately from past experiences of seeing
'big things done with JS'.  Does couchbase come with a facility for a JSON
import?

**Comment by nwalters512** [comment]
The desktop app is used to...
* Pull an event list off TBA to select the event we're at
* Load team/match data for our event
* Configure the set of people that can log into the app to scout; this includes setting up our rudimentary permissions system
* Add pictures to our dataset

The mobile app does use couchbase. That's necessary based on the current synchronization scheme: on the tablet, we open both the on-tablet and on-flash-drive databases at the same time and then crudely force them to be consistent with each other.

I'm already anticipating at least some amount of custom data processing. Specifically, the nedb library I mentioned runs in append-only mode for performance reasons, so any modifications to the database are appended to the end of the file, and whenever the file is read from disk, all those changes are compacted back into existing records. Obviously bad if you want to scale to millions of records, but benchmarks looked good for a small dataset like we'd have.

A checksum would definitely help us check for consistency, but wouldn't do much to help us recover from an inconsistent state. The idea of zipping everything up into a file and sending it over is probably the best option: let the filesystem handle consistency for us.

**Comment by smitty6791** [comment]
Is there any reason that we don't sync this over USB like every other
PC/app out there?  That would take the USB OTG problem out of the
equation(no more using the port as a host port).  It would also remove the
database storage differences. You could use the native sqlite database on
the tablet.   Each tablet would need to be plugged into the PC to sync it,
but you don't need to worry about a common format of a file anymore.  The
format differences are handled by the sync messages.

**Comment by garwards** [comment]
Does sqlite have a sync/replicate function built in?

**Comment by nwalters512** [comment]
I guess I'll specify exactly the problem domain here and what type of solution we need.

Our issue is that we need a data storage system whose binary format is consistent between both platforms and languages due to the architecture I chose/was forced into by the fact that we can't use any sort of wireless communications at event: a file on a physical disk serves as the means of data synchronization and central source of truth. I don't know of any databases where binaries are guaranteed to be consistent between platforms and languages (maybe you do?).

Couchbase happened to work for us because the Java client library uses SQLite as the data storage layer, so both the Java desktop app and Android (Java) scouting app could read the same binary file. But that's not the case with any other language; in any case, we're forced to interact with the underlying database through a Couchbase client library, so if one doesn't exist for a certain platform/language combo, we're out of luck.

The way I see it, we're kind of forced into rolling our own solution, and I don't necessarily know enough to write anything other than a pretty naive plaintext storage system like nedb uses.

**Comment by nwalters512** [comment]
We don't use a cable for the purpose of making data easier to shuttle around. At champs, for instance, it's pretty easy to send a freshman with a flash drive to the pit - slightly less easy to get someone to sacrifice a laptop for the weekend and have that laptop be run all over the place.

**Comment by nwalters512** [comment]
SQLite doesn't have built-in syncing, but Couchbase does; it all runs over http. If we're willing to devote a computer to running a Couchbase server and can make http work over USB, that's not a bad option.

**Comment by smitty6791** [comment]
Would a 2 tablet scheme solve this.  The drive coach(let's just call him
Steve for short) has the tablet for the current match.  A second table is
being loaded for the next match.  The runner takes the new tablet down,
switches with Steve and runs the first tablet up for reload of the next
match.  Honestly, that seems to be more of what I saw last year.

Either way, it really doesn't matter which way we sync these.  If the USB
drive is easier to deal with, that's fine, but it looks like you may have
to home brew the protocol.  Which lets face it, you'll have to do either
way.  It just depends on the amount of data marshaling you need to do.
Unfortunately, I don't have a good answer is to how much that really is.

**Comment by smitty6791** [comment]
We would have to check, but I recall correctly, the USB device connection
essentially creates a network device on the tablet and the PC side also
gets a network device.  Now you have a standard network connection.  Most
development boards to the same thing these days.  The beaglebone, the pi,
they all create a network interface.  That gives you access over Http.  Now
we would just need to run a server on the PC.  If we had say...a PC running
linux it becomes even easier to run a webserver.  You can do it on windows,
but its just more of a pain.  I just though of something ...we have a
chromebook from some drawing at a show..  That would be android to
android.  And you can write both ends as native android apps.  Would that
help out?

**Comment by garwards** [comment]
Similarly, an RPi in a box with a battery would be small and act as a
server of sorts.  It can run headless at comps as all interaction would be
from the app making requests, or a laptop doing much the same.  Instant
portable mini server.

**Comment by smitty6791** [comment]
Good point!  If you still want android on it, there is a port available as
well.  Lots of possibilities with that one.  Heck, we could run that off an
old robot battery if necessary.

**Comment by nwalters512** [comment]
That's actually a really great idea. You guys could design a nice 3D printed case and battery circuitry and release the whole thing alongside WildRank so that other teams can still use it; I know at least 876 is pretty invested in our software. We could wrap up the whole server in a container (Docker? Yay buzzwords!) to make deploying idiotproof and let people run on hardware of their choice. I think I'll have to get my RPi out of the closet and start playing around with that.

**Comment by nwalters512** [comment]
Update: looks like Android doesn't automatically establish a network connection when plugged in. To get syncing to work, we'd have to use adb to set up reverse port forwarding from a port on the Android device to the local port that the server is running on. That would mean detecting when an Android device is mounted and automatically setting up port forwarding then. I'm guessing the answer is "it's Linux, duh", but is there a way to trigger something when a device is mounted? Or do we just have to poll?

**Comment by garwards** [comment]
Yeah you can run a script when it connects a USB device.  Here's an example:
http://unix.stackexchange.com/questions/65891/how-to-execute-a-shellscript-when-i-plug-in-a-usb-device

**Comment by nwalters512** [comment]
Update, part 2: Couchbase only seems to support 2-way replication over HTTP in a setup of Couchbase Server<--->Couchbase Lite. We also can't use any default API to access data on the server end, meaning we'd have to build an API on top of the underlying server.

That would give us 2 options:

### Option 1
Run a bare Couchbase server on an RPi; keep the existing Java desktop app; use built-in replication to sync everything.

__Pros__: don't have to do manual data merging and conflict resolution. Yay!
__Cons__: not much significant gain with a lot of added complexity in the form of the RPi.

### Option 2
Build an API on top of the Couchbase server; use whatever we want to sync with it.

__Pros__: lots more flexibility with what we can use to develop apps. We just have to make sure we have an easy way of fowarding ports to the RPi.
__Cons__: still added compexity in the form of the RPi. Still have to do manual data syncing.
  
<br />
<br />

After doing some more research, I came across this gem on http://sqlite.org/different.html:


>__Stable Cross-Platform Database File__
>
>The SQLite file format is cross-platform. A database file written on one machine can be copied to and used on a different machine with a different architecture. Big-endian or little-endian, 32-bit or 64-bit does not matter. All machines use the same file format. Furthermore, the developers have pledged to keep the file format stable and backwards compatible, so newer versions of SQLite can read and write older database files.
>
>Most other SQL database engines require you to dump and restore the database when moving from one platform to another and often when upgrading to a newer version of the software.

So, that actually makes an SQLite database a really good candidate for a cross-platform data storage solution.

__Pros__: consistent binary format! If a library exists for a platform/language combo, we can use it.
__Cons__: We still have to do manual synchronization; this may be (much) more difficult in SQL given that it's significantly more low-level than Couchbase. No native support for JSON-like documents (unlike Couchbase), so we have to take extra care when dealing with those; most data we collect is stored as hierarchical documents.
